### DISCUSSION TITLE
Trailing slash rewrite - duplicated content SEO issue

### DISCUSSION THREAD

**Discussion by sinedoOo** [nginx/unit#1035] [Support] [answered]
Hello, I'm wondering how we can achieve trailing slash rewrite in Nginx Unit in case of PHP (Laravel) app? I need to return 301 redirect for all requests that ends with `/` to address without it. The purpose of it is to fight with duplicated content from SEO perspective.

In Nginx it can be achieved with one simple rewrite:

`rewrite ^/(.*)/$ /$1 permanent;` 

How we can do it in case of Nginx Unit?

**Comment by lcrilly** [comment, marked as answer]
This routes array should do what you want

```json
[
  {
    "match": {
      "uri": "/*/"
    },
    "action": {
      "rewrite": "`${uri.slice(0,-1)}`",
      "return": 301,
      "location": "$uri"
    }
  },
  {
    "action": {
      "return": 204
    }
  }
]
```

```
$ curl -i http://localhost/foo/bar/
HTTP/1.1 301 Moved Permanently
Location: /foo/bar
Server: Unit/1.31.1
Date: Mon, 18 Dec 2023 10:44:12 GMT
Content-Length: 0
```

**Comment by sinedoOo** [comment]
First of all thank you for super fast reply. I tried this config, but it doesn't work :/ I'm receiving 301 redirect, but with broken URL.

```
curl -i http://localhost/foo/bar/
HTTP/1.1 301 Moved Permanently
Location: %60$%7Buri.slice(0,-1)%7D%60
Server: Unit/1.31.1
Date: Mon, 18 Dec 2023 13:36:48 GMT
Content-Length: 0
```

**Reply by lcrilly** [reply]
Perhaps you don't have the JavaScript module enabled? The backtick and braces are appearing as literal characters and then URL-encoded.

Please check `unitd --version` for the `--njs` configure option. This is the default on all of the official builds/packages but not if you compiled from source.

If you're building from source then follow the njs guide here
https://unit.nginx.org/howto/source/#source-njs

**Reply by sinedoOo** [reply]
I'm using the build provided by REMI repo in order to have latest PHP version.

```
unit version: 1.31.1
configured as ./configure --libdir=/usr/lib64 --prefix=/usr --statedir=/var/lib/unit --control=unix:/run/unit/control.sock --pid=/run/unit/unit.pid --runstatedir=/var/run --log=/var/log/unit/unit.log --logdir=/var/log --tmpdir=/var/tmp --user=unit --group=unit --openssl --cc-opt='-O2 -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -march=x86-64-v2 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection' --tests --modulesdir=/usr/lib64/unit/modules
```

So basically I have to move to official repo? But from my understanding it's tied to PHP version from official RHEL repo. Correct? Or is there any option to do it differently?

**Reply by sinedoOo** [reply]
I solved this issue. Now I use `unit` package from official repo, and only `unit-php` from Remi repo. In order to achieve it, I had to add `exclude=unit` to remi-safe repo file.

Redirect works now, thank you for help @lcrilly!

**Reply by lcrilly** [reply]
Great news! 🎉 

Perhaps we should ask @remicollet to include **njs** in the Unit-PHP packages 💡

**Reply by sinedoOo** [reply]
I already did it, but he explained why it's not possible. Please check the message on official Remi's forum: https://forum.remirepo.net/viewtopic.php?id=4493

**Reply by lcrilly** [reply]
Thanks for that. There might be some confusion by the term "module" because we support Node.JS as a language runtime and "NGINX JavaScript" (njs) as a configuration extension to the Unit core.

We'll discuss with Remi in more detail because njs enables many use cases independent of the "language module".

**Reply by ac000** [reply]
Looks like there's been some [movement](https://forum.remirepo.net/viewtopic.php?pid=12826#p12826) from Remi's side...

**Reply by sinedoOo** [reply]
I just tested it and it works now perfectly fine directly from Remi repo.
